### PR TITLE
clang: skip the `-nostdlibinc` patch on Darwin; ld64: search standard library locations

### PR DESCRIPTION
**Pull request by emilazy**
It kind of feels like we should find a better way of doing this on Linux too, but I’m picking the conservative option for now. Not tested in any way yet.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by alyssais — COMMENTED** [review]
Shouldn't this be targetPlatform? When cross compiling from Darwin to Linux, we'd want the same behaviour as building natively on Linux, right?

**Comment by emilazy** [comment]
Yyyyes, maybe. Except the reasoning for why it’s safe and won’t cause things like https://github.com/NixOS/nixpkgs/issues/151879 is based on properties of Darwin as a host platform. So I don’t know, really. Part of me hopes that when cross to Darwin is a thing we’ll have figured out a way to get rid of this patch on Linux too, but I can change it to `targetPlatform` if you think it’s best, since after all there is no way to hurt yourself with it currently anyway.

**Comment by emilazy** [comment]
Oh, sorry, I misread you as talking about cross the other way around. I think it is safe for cross to Linux. This patch is only here to stop host platform stuff leaking in on non‐NixOS Linux. But it can’t really leak in on macOS because the stuff that would leak in isn’t there. And you’d have a cross sysroot for that case, anyway. But I guess the compiler would behave slightly differently (i.e., better, basically).

File: pkgs/development/compilers/llvm/common/clang/default.nix [hunk 1]
```diff
@@ -81,13 +81,17 @@ let
...
+      # See the comment on the `add-nostdlibinc-flag.patch` patch in
+      # `../default.nix` for why we skip Darwin here.
+      if lib.versionOlder release_version "13" && !stdenv.hostPlatform.isDarwin then ''
+        sed -i -e 's/DriverArgs.hasArg(options::OPT_nostdlibinc)/true/' \
+               -e 's/Args.hasArg(options::OPT_nostdlibinc)/true/' \
+               lib/Driver/ToolChains/*.cpp
+      '' else ''
+        (cd tools && ln -s ../../clang-tools-extra extra)
+      ''
+    ) + lib.optionalString stdenv.hostPlatform.isMusl ''
```
Comment (paparodeo) [inline pkgs/development/compilers/llvm/common/clang/default.nix:94]: this is removed in https://github.com/NixOS/nixpkgs/pull/348568

Comment (sternenseemann) [reply pkgs/development/compilers/llvm/common/clang/default.nix:94]: Yes, https://github.com/NixOS/nixpkgs/pull/348568/commits/87086972d65c6a89b5a43182d6e845a80f807bce.

Comment (reckenrode) [reply pkgs/development/compilers/llvm/common/clang/default.nix:94]: It’s replaced with a patch, but I think the logic of not using it on Darwin would remain in that case.

**Comment by reckenrode** [comment]
This is the PR that added the patch. It fixed a build failure on Linux. I tried looking through the history, but this behavior has been around for quite a while. I also assume it’s purity related, but I don’t know.

https://github.com/NixOS/nixpkgs/issues/151879

**Comment by Ericson2314** [comment]
The name of the patch seems confusing? It implies is adding a flag, which should be harmless? But then the comment says something else.

**Comment by emilazy** [comment]
So the situation is:

1. This reasoning for why this patch is harmful and requires annoying workarounds in downstream packages applies to *targeting* macOS.

2. The reasoning for why this patch is mostly doesn’t prevent any brokenness applies to *building* on macOS.

3. Theoretically, this change is backwards‐incompatible in very marginal cases where you’re running an unsandboxed build on macOS without any SDK and have Homebrew headers installed. One could construe this as a breaking change to a Release Critical Package, although I think the circumstances in which it would arise are very much not release‐critical.

Ideally, the patch should be applied based on `targetPlatform`. However, that would expose a hypothetical Linux‐to‐Darwin cross toolchain to regressions of the bugs Randy linked. I don’t know whether we care about that, especially as such toolchains don’t currently exist. Perhaps it could apply when both the host and target platform are Darwin for now to be conservative? I will go with whatever people think is best.

I would like to get this in for 24.11 because currently Rust bindgen stuff is broken on macOS and I don’t think it’s meaningfully backwards‐incompatible in a way that anyone will notice, but pinging @RossComputerGuy just in case.

**Comment by emilazy** [comment]
Pushed to clean up now‐unnecessary hacks in the Darwin stdenv, and apply the same change to `ld64` as @reckenrode pointed out that the same issues came up there with things like the .NET build. Will get some builds run to test that I didn’t break anything here.

**Comment by emilazy** [comment]
Successfully built up to `hello` and @reckenrode and @niklaskorz are testing the .NET build and Rust bindgen stuff. Ran into an issue with Swift; had to undo an equivalent obsolete patch there. Will confirm that fixed it tomorrow (though Swift doesn’t even build on Hydra right now).

**Review by RossComputerGuy — APPROVED** [review]
LGTM before the 25th.

**Review by niklaskorz — APPROVED** [review]
zed-editor now builds fine on darwin without the `NIX_CFLAGS_COMPILE = "-F${apple-sdk_15.sdkroot}/System/Library/Frameworks";` workaround for rust-bindgen 🥳 

```
┏━ Dependency Graph:
┃    ┌─ ✔ blade-util-0.1.0 ⏱ 1s
┃    ├─ ✔ blade-macros-0.3.0 ⏱ 1s
┃    ├─ ✔ tree-sitter-gowork-0.0.1
┃    ├─ ✔ tree-sitter-heex-0.0.1
┃    ├─ ✔ xim-ctext-0.3.0
┃    ├─ ✔ tree-sitter-md-0.3.2
┃    ├─ ✔ xim-0.4.0
┃    ├─ ✔ xim-parser-0.2.1
┃    ├─ ✔ xkbcommon-0.7.0
┃    ├─ ✔ tree-sitter-yaml-0.6.1 ⏱ 8s
┃ ┌─ ✔ cargo-vendor-dir ⏱ 34s
┃ ├─ ✔ cargo-bundle-unstable-2023-08-18 ⏱ 2m46s
┃ ├─ ✔ cargo-about-0.6.4 ⏱ 3m47s
┃ ✔ zed-editor-0.157.5 ⏱ 41m5s
┣━━━ Builds
┗━ ∑ ⏵ 0 │ ✔ 1903 │ ⏸ 0 │ Finished at 08:30:05 after 11h24m32s
```

**Comment by emilazy** [comment]
Zed working confirms that this PR accomplishes its primary goal 🎉 Still waiting to hear about .NET. Swift is being a pain and we might have to back out those changes for now. Testing dropping the redundant `-isysroot` to fix https://github.com/NixOS/nixpkgs/pull/350853.

**Comment by alyssais** [comment]
Ack.

**Comment by emilazy** [comment]
Dropped the Swift changes; it’s too haunted to deal with right now and it works okay without the commit. Moved the flags for backwards‐compatibility with the current bootstrap tools into the `stdenv`. This should be ready to go pending confirmation that it does in fact build .NET successfully.

**Comment by emilazy** [comment]
Summary of the state of things: it bootstraps fine and fixes the regression, can build Swift, Zed, etc., but the .NET build fails due to Swift‐related `ld64` wrapper flag issues that need to be diagnosed; it’s possible that part of the flags wrapper just needs to be restored, at least for now, though it could also theoretically be worked around in the .NET rebuild. I will investigate tomorrow; I hope it is okay for this to miss the deadline by a day since the circumstances in which this is at all a breaking change are very marginal and it will be many days before the `staging` cycle this lands in even starts anyway.

**Comment by emilazy** [comment]
It turns out that .NET’s build system just has `-L/usr/lib/swift` references that our awful CMake hook was obliterating and we were inadvertently working around that, and everything about our toolchain is actually working as expected with this PR. We may want to consider automatically adding `usr/lib/swift` to match the closed‐source `ld-prime` in the future, but that’s out of scope for this. It’s less than an hour late so hopefully this is okay to merge.

**Pull request merged**